listenbrainz/mbid_mapping_writer/mbid_mapper.py

```diff
@@ -0,0 +1,284 @@
+MATCH_TYPE_HIGH_QUALITY = 3
+MATCH_TYPE_EXACT_MATCH = 4
+
+MATCH_TYPE_HIGH_QUALITY_MAX_EDIT_DISTANCE = 2
+MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE = 5
+
+ENGLISH_STOP_WORD_INDEX = {k: 1 for k in ENGLISH_STOP_WORDS}
+
+
+def prepare_query(text):
+    return unidecode(re.sub(" +", " ", re.sub(r'[^\w ]+', '', text)).strip().lower())
+
+
+class MBIDMapper():
+    """
+        This class performs a lookup of one or more artist credit name and recording name pairs
+        to find the best possible match in MusicBrainz. This query will unaccent query
+        parameters, be resistant to typos and attempt to modify the search parameters
+        in effort to find a best match.
+
+        For instance, if "Glory Box feat. Sloppy Jo" isn't matched, it will try to match "Glory Box".
+        It is the caller's responsibility to ensure that the results meet the criteria for the
+        given task. This mapping lookup aims to find a match by a number of means necessary that
+        may or may not be acceptable to the caller.
+
+        Other possible detunings:
+        - Swap artist/recording
+        - remove track numbers from recording
+
+        Overall these detunings and edit distances will need more tuning as we start using this
+        endpoint for our MessyBrainz mapping. For this reason the debug flag and optional
+        debug output remains since more debugging/tuning will need to be done later on.
+    """
+
+    MATCH_TYPE_HIGH_QUALITY_MAX_EDIT_DISTANCE = 2
+    MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE = 5
+
+    def __init__(self, timeout=DEFAULT_TIMEOUT, remove_stop_words=False, debug=False):
+        self.debug = debug
+        self.log = []
+
+        self.client = typesense.Client({
+            'nodes': [{
+                'host': config.TYPESENSE_HOST,
+                'port': config.TYPESENSE_PORT,
+                'protocol': 'http',
+            }],
+            'api_key': config.TYPESENSE_API_KEY,
+            'connection_timeout_seconds': timeout
+        })
+        self.remove_stop_words = remove_stop_words
+        self.is_ac_detuned = False
+        self.is_r_detuned = False
+
+    def _log(self, str):
+        if self.debug:
+            self.log.append(str)
+
+    def read_log(self):
+        log = self.log
+        self.log = []
+
+        return log
+
+    def detune_query_string(self, query, is_artist_credit):
+        """
+            Detune (remove known extra cruft) a metadata field. If a known
+            trailing string exists in the input string, that string and everything
+            after it is removed.
+        """
+
+        strings = ["(", "[", " ft ", " ft. ", " feat ", " feat. ", " featuring ", " - "]
+        if is_artist_credit:
+            strings.insert(0, ",")
+            strings.append(" with ")
+
+        for s in strings:
+            index = query.find(s)
+            if index >= 0:
+                return query[:index].strip()
+
+        # Yes, this is actually correct.
```

Comment:
The documentation of this method says that it removes extra cruft so I would expect that if extra cruft does not exist the same string is returned while actually "" is returned. I think it makes sense to:
1) To mention this in the docstring of the method
2) Explain the comment here - why is this correct?

Reply:
With some effort, I think I was able to figure out why this is correct. My understanding is that L255 later in this file does not execute if there's nothing to detune.